#This analysis is intended to compare siScramble vs. siCIC treatment for two cell lines,
#X1 and CDS2 (X292), which are both patient-derived CIC-DUX4 tumor cell lines.
#GeneCounts were generated by STAR during alignment to GRCh38.
#A simple edgeR analysis using the classic pipeline and the exact test is used to compare siScamble & siCIC for each cell line individually.
#Since the comparisons are done separately one cell line at a time, the log(cpm) data for each cell line are normalized to themselves only, and should...
#not be compared with each other (i.e. do not compare X1 to CDS2 log(cpm)).
#Other comparisons with ChIP-seq data are also performed.
#Contact: cuyler.luck@ucsf.edu or ross.okimoto@ucsf.edu

library(data.table)
library(ggbiplot)
library(dplyr)
library(gmodels)
library(biomaRt)
library(edgeR)
library(tidyr)
library(ggpubr)

#set wd as appropriate
setwd("/Volumes/cuyler/ucsf_okimoto_lab/nicole_RNAseq/hg38_data")

#load the data, skipping the first 4 lines which just provide information on mapping
X1_scrm1 = fread("ReadsPerGene/X1_siSCRMBL1ReadsPerGene.out.tab", skip = 4)
X1_scrm2 = fread("ReadsPerGene/X1_siSCRMBL2ReadsPerGene.out.tab", skip = 4)
X1_cic1 = fread("ReadsPerGene/X1_siCIC1ReadsPerGene.out.tab", skip = 4)
X1_cic2 = fread("ReadsPerGene/X1_siCIC2ReadsPerGene.out.tab", skip = 4)
CDS2_scrm1 = fread("ReadsPerGene/X292_SCRMBL1ReadsPerGene.out.tab", skip = 4)
CDS2_scrm2 = fread("ReadsPerGene/X292_SCRMBL2ReadsPerGene.out.tab", skip = 4)
CDS2_cic1 = fread("ReadsPerGene/X292_siCIC1ReadsPerGene.out.tab", skip = 4)
CDS2_cic2 = fread("ReadsPerGene/X292_siCIC2ReadsPerGene.out.tab", skip = 4)

#this is unstranded library prep, so we can simplify the dataframes to just columns 1 and 2.
X1_scrm1 = dplyr::select(X1_scrm1, c(1,2))
X1_scrm2 = dplyr::select(X1_scrm2, c(1,2))
X1_cic1 = dplyr::select(X1_cic1, c(1,2))
X1_cic2 = dplyr::select(X1_cic2, c(1,2))
CDS2_scrm1 = dplyr::select(CDS2_scrm1, c(1,2))
CDS2_scrm2 = dplyr::select(CDS2_scrm2, c(1,2))
CDS2_cic1 = dplyr::select(CDS2_cic1, c(1,2))
CDS2_cic2 = dplyr::select(CDS2_cic2, c(1,2))

#then, we can combine all of these into one dataframe for visualization. change the column names first so we know what's what.
colnames(X1_scrm1) = c("Gene", "X1_scrm1")
colnames(X1_scrm2) = c("Gene", "X1_scrm2")
colnames(X1_cic1) = c("Gene", "X1_cic1")
colnames(X1_cic2) = c("Gene", "X1_cic2")
colnames(CDS2_scrm1) = c("Gene", "CDS2_scrm1")
colnames(CDS2_scrm2) = c("Gene", "CDS2_scrm2")
colnames(CDS2_cic1) = c("Gene", "CDS2_cic1")
colnames(CDS2_cic2) = c("Gene", "CDS2_cic2")

master = inner_join(X1_scrm1, X1_scrm2, by = "Gene")
master = inner_join(master, X1_cic1, by = "Gene")
master = inner_join(master, X1_cic2, by = "Gene")
master = inner_join(master, CDS2_scrm1, by = "Gene")
master = inner_join(master, CDS2_scrm2, by = "Gene")
master = inner_join(master, CDS2_cic1, by = "Gene")
master = inner_join(master, CDS2_cic2, by = "Gene")

master = tibble::column_to_rownames(master, "Gene")

#now let's run an edgeR standard pipeline, one at a time, for the two cell lines. comparing siScramble vs. siCIC.
#we will be able to look at clustering plots later, within cell line, using edgeR (TMM) normalized CPM data.

### X1 analysis

X1_only = dplyr::select(master, X1_scrm1, X1_scrm2, X1_cic1, X1_cic2)
X1_groups = c("scrm", "scrm", "siCIC", "siCIC")
x1 = DGEList(counts = X1_only, group = X1_groups)
x1_keep = filterByExpr(x1)
x1 = x1[x1_keep, , keep.lib.sizes = F]
x1 = calcNormFactors(x1)
x1 = estimateDisp(x1)
x1_et = exactTest(x1)
x1_results = x1_et$table
x1_results = tibble::rownames_to_column(x1_results, "ENSG_ID")
#adjust p values using FDR method
x1_results = dplyr::mutate(x1_results, p_fdr_adj = p.adjust(PValue, method = "fdr"))
#and make a -log10(q) for use in volcano plot
x1_results = dplyr::mutate(x1_results, neglog10_p_fdr_adj = -log10(p_fdr_adj))


#Translate gene names from ENSG IDs to gene symbols using biomaRt
#Using Ensembl genes for GRCh38.p13.
#This will almost definitely lose some genes for which multiple symbols match or for which no symbols exist

#ensembl = useEnsembl(biomart = "ensembl")
#listDatasets(ensembl) #this shows that at the time of analysis, the 'hsapiens_gene_ensembl' dataset is for GRCh38.p13 which is the same version I used to align to.
ensembl = useEnsembl(biomart="ensembl", dataset='hsapiens_gene_ensembl', mirror = "uswest")

#listFilters(ensembl) #I will filter by ensembl_gene_id
#listAttributes(ensembl) #I will retrieve hgnc_symbol and ensembl_gene_id
#actually get the key from biomaRt
symbol_key = getBM(attributes=c("hgnc_symbol","ensembl_gene_id"), filters="ensembl_gene_id",values=x1_results$ENSG_ID, mart=ensembl)
#trim out entries that have no gene symbol
symbol_key = symbol_key[symbol_key$hgnc_symbol!="",]

#make sure there are unique matches
length(symbol_key$hgnc_symbol)
length(unique(symbol_key$hgnc_symbol))
length(unique(symbol_key$ensembl_gene_id))
#there are duplicate symbols, let's get rid of them
symbol_key = symbol_key[!duplicated(symbol_key$hgnc_symbol),]
#check again
length(symbol_key$hgnc_symbol)
length(unique(symbol_key$hgnc_symbol))
length(unique(symbol_key$ensembl_gene_id))
#now all lengths match, good.

#make a new X1 DE results table that just includes genes that have symbols.
colnames(symbol_key) = c("symbol","ENSG_ID")
x1_symbol_results = left_join(symbol_key, x1_results, by="ENSG_ID")
#this loses about 1153 genes for which there were ENSG IDs but which did not map nicely to gene symbols.

#write this to a file so we can look at it other ways (and send it to inquiring minds... like the boss)
write.table(x1_symbol_results, file = "X1_edgeR_DE.tsv", sep = "\t", row.names = F, quote = F)

#read in Watson high-confidence CIC-DUX4 genes (Watson et al bioRxiv 2019)
watson = fread("watson_biorxiv_genes.tsv", header = F)

#plot!

#These cutoffs are:
# -log10(q) value line is drawn at 4, which is 0.0001
# log2FC cutoff lines are drawn at +/- 0.58, which is approx. 0.66 & 1.5 FC

#this plot colors CIC red and the genes from Watson gold
pdf(file = "plots/X1.watsongenes.pdf", width = 4, height = 4)
ggplot(x1_symbol_results, aes(x = logFC, y = neglog10_p_fdr_adj)) + 
  geom_hline(yintercept = 4, linetype = 2) +
  geom_vline(xintercept = 0.58, linetype = 2) +
  geom_vline(xintercept = -0.58, linetype = 2) +
  geom_point(alpha = 0.15, fill = "black", pch=21, size = 3) +
  theme_classic() +
  geom_point(data = x1_symbol_results[x1_symbol_results$symbol == "CIC",], fill = "red", pch=21, size = 3) +
  geom_point(data = x1_symbol_results[x1_symbol_results$symbol %in% watson$V1,], fill = "gold", pch=21, size = 3) +
  xlab("log2(FC)")+
  ylab("-log10(q)")+
  ggtitle("X1, CIC (red) and Watson genes (red)")
dev.off()

#this plot colors CIC red, ETV1/4/5 light blue, and DUSP6 regular blue
pdf(file = "plots/X1.PEA3.DUSP6.pdf", width = 4, height = 4)
ggplot(x1_symbol_results, aes(x = logFC, y = neglog10_p_fdr_adj)) + 
  geom_hline(yintercept = 4, linetype = 2) +
  geom_vline(xintercept = 0.58, linetype = 2) +
  geom_vline(xintercept = -0.58, linetype = 2) +
  geom_point(alpha = 0.15, fill = "black", pch=21, size = 3) +
  theme_classic() +
  geom_point(data = x1_symbol_results[x1_symbol_results$symbol == "CIC",], fill = "red", pch=21, size = 3) +
  geom_point(data = x1_symbol_results[x1_symbol_results$symbol %in% c("ETV1", "ETV4", "ETV5"),], fill = "light blue", pch=21, size = 3) +
  geom_point(data = x1_symbol_results[x1_symbol_results$symbol %in% c("DUSP6"),], fill = "blue", pch=21, size = 3) +
  xlab("log2(FC)")+
  ylab("-log10(q)")+
  ggtitle("X1, CIC (red), ETV1/4/5 (light blue), and DUSP6 (blue)")
dev.off()


#this plot colors CIC red and SPRED/SPRY family members light green
pdf(file = "plots/X1.SPRY.SPRED.pdf", width = 4, height = 4)
ggplot(x1_symbol_results, aes(x = logFC, y = neglog10_p_fdr_adj)) + 
  geom_hline(yintercept = 4, linetype = 2) +
  geom_vline(xintercept = 0.58, linetype = 2) +
  geom_vline(xintercept = -0.58, linetype = 2) +
  geom_point(alpha = 0.15, fill = "black", pch=21, size = 3) +
  theme_classic() +
  geom_point(data = x1_symbol_results[x1_symbol_results$symbol == "CIC",], fill = "red", pch=21, size = 3) +
  geom_point(data = x1_symbol_results[x1_symbol_results$symbol %in% c("SPRY1", "SPRY2", "SPRY3", "SPRY4"),], fill = "green3", pch=21, size = 3) +
  geom_point(data = x1_symbol_results[x1_symbol_results$symbol %in% c("SPRED1", "SPRED2", "SPRED3"),], fill = "palegreen2", pch=21, size = 3) +
  xlab("log2(FC)")+
  ylab("-log10(q)")+
  ggtitle("X1, CIC (red), SPRY1/2/3/4 (darker green), SPRED1/2/3 (lighter green)")
dev.off()

#this plot colors CIC red, POLE pink, and CDH tan
pdf(file = "plots/X1.POLE.CDH.pdf", width = 4, height = 4)
ggplot(x1_symbol_results, aes(x = logFC, y = neglog10_p_fdr_adj)) + 
  geom_hline(yintercept = 4, linetype = 2) +
  geom_vline(xintercept = 0.58, linetype = 2) +
  geom_vline(xintercept = -0.58, linetype = 2) +
  geom_point(alpha = 0.15, fill = "black", pch=21, size = 3) +
  theme_classic() +
  geom_point(data = x1_symbol_results[x1_symbol_results$symbol == "CIC",], fill = "red", pch=21, size = 3) +
  geom_point(data = x1_symbol_results[x1_symbol_results$symbol %in% c("POLE"),], fill = "pink", pch=21, size = 3) +
  geom_point(data = x1_symbol_results[x1_symbol_results$symbol %in% c("CDH4"),], fill = "tan", pch=21, size = 3) +
  xlab("log2(FC)")+
  ylab("-log10(q)")+
  ggtitle("X1, CIC (red), POLE (pink), CDH4 (tan)")
dev.off()

#extract log CPM values for dimensionality reduction and other things that require normalized data.
x1_logcpm = cpm(x1, log=TRUE)
x1_logcpm_df = as.data.frame(x1_logcpm)
x1_logcpm_df = tibble::rownames_to_column(x1_logcpm_df, "ENSG_ID")

#visualize by PCA using logCPM values
x1_pca_obj = fast.prcomp(t(x1_logcpm), center=T, scale=T)
pdf(file = "plots/X1_logCPM_PCA.pdf", width = 3, height = 3)
ggbiplot(x1_pca_obj, var.axes = F, groups = c("siSCRM", "siSCRM", "siCIC", "siCIC")) +
  theme_classic()
dev.off()




### CDS2 analysis
CDS2_only = dplyr::select(master, CDS2_scrm1, CDS2_scrm2, CDS2_cic1, CDS2_cic2)
CDS2_groups = c("scrm", "scrm", "siCIC", "siCIC")
cds2 = DGEList(counts = CDS2_only, group = CDS2_groups)
cds2_keep = filterByExpr(cds2)
cds2 = cds2[cds2_keep, , keep.lib.sizes = F]
cds2 = calcNormFactors(cds2)
cds2 = estimateDisp(cds2)
cds2_et = exactTest(cds2)
cds2_results = cds2_et$table
cds2_results = tibble::rownames_to_column(cds2_results, "ENSG_ID")
#adjust p values using FDR method
cds2_results = dplyr::mutate(cds2_results, p_fdr_adj = p.adjust(PValue, method = "fdr"))
#and make a -log10(q) for use in volcano
cds2_results = dplyr::mutate(cds2_results, neglog10_p_fdr_adj = -log10(p_fdr_adj))

#Translate gene names from ENSG IDs to gene symbols using biomaRt
#much of the legwork is done above to prepare the biomaRt call
#similar except we change the values to ask for
cds2_symbol_key = getBM(attributes=c("hgnc_symbol","ensembl_gene_id"), filters="ensembl_gene_id",values=cds2_results$ENSG_ID, mart=ensembl)
#trim out entries that have no gene symbol
cds2_symbol_key = cds2_symbol_key[cds2_symbol_key$hgnc_symbol!="",]

#make sure there are unique matches
length(cds2_symbol_key$hgnc_symbol)
length(unique(cds2_symbol_key$hgnc_symbol))
length(unique(cds2_symbol_key$ensembl_gene_id))
#there are duplicate symbols, let's get rid of them
cds2_symbol_key = cds2_symbol_key[!duplicated(cds2_symbol_key$hgnc_symbol),]
#check again
length(cds2_symbol_key$hgnc_symbol)
length(unique(cds2_symbol_key$hgnc_symbol))
length(unique(cds2_symbol_key$ensembl_gene_id))
#now all lengths match, good.

#merge with results
colnames(cds2_symbol_key) = c("symbol", "ENSG_ID")
cds2_symbol_results = left_join(cds2_symbol_key, cds2_results, by="ENSG_ID")

#write this to a file so we can look at it other ways (and send it to inquiring minds... like the boss)
write.table(cds2_symbol_results, file = "CDS2_edgeR_DE.tsv", sep = "\t", row.names = F, quote = F)


#plot!

#These cutoffs are:
# -log10(q) value line is drawn at 4, which is 0.0001
# log2FC cutoff lines are drawn at +/0 0.58, which is approx. 0.66 & 1.5 FC

#this plot colors CIC red and the genes from Watson gold
pdf(file = "plots/CDS2.watsongenes.pdf", width = 4, height = 4)
ggplot(cds2_symbol_results, aes(x = logFC, y = neglog10_p_fdr_adj)) + 
  geom_hline(yintercept = 4, linetype = 2) +
  geom_vline(xintercept = 0.58, linetype = 2) +
  geom_vline(xintercept = -0.58, linetype = 2) +
  geom_point(alpha = 0.15, fill = "black", pch=21, size = 3) +
  theme_classic() +
  geom_point(data = cds2_symbol_results[cds2_symbol_results$symbol == "CIC",], fill = "red", pch=21, size = 3) +
  geom_point(data = cds2_symbol_results[cds2_symbol_results$symbol %in% watson$V1,], fill = "gold", pch=21, size = 3) +
  xlab("log2(FC)")+
  ylab("-log10(q)")+
  ggtitle("CDS2, CIC (red) and Watson genes (red)")
dev.off()

#this plot colors CIC red, ETV1/4/5 light blue, and DUSP6 regular blue
pdf(file = "plots/CDS2.PEA3.DUSP6.pdf", width = 4, height = 4)
ggplot(cds2_symbol_results, aes(x = logFC, y = neglog10_p_fdr_adj)) + 
  geom_hline(yintercept = 4, linetype = 2) +
  geom_vline(xintercept = 0.58, linetype = 2) +
  geom_vline(xintercept = -0.58, linetype = 2) +
  geom_point(alpha = 0.15, fill = "black", pch=21, size = 3) +
  theme_classic() +
  geom_point(data = cds2_symbol_results[cds2_symbol_results$symbol == "CIC",], fill = "red", pch=21, size = 3) +
  geom_point(data = cds2_symbol_results[cds2_symbol_results$symbol %in% c("ETV1", "ETV4", "ETV5"),], fill = "light blue", pch=21, size = 3) +
  geom_point(data = cds2_symbol_results[cds2_symbol_results$symbol == "DUSP6",], fill = "blue", pch=21, size = 3) +
  xlab("log2(FC)")+
  ylab("-log10(q)")+
  ggtitle("CDS2, CIC (red), ETV1/4/5 (light blue), and DUSP6 (blue)")
dev.off()

#this plot colors CIC red and SPRED/SPRY family members light green
pdf(file = "plots/CDS2.SPRY.SPRED.pdf", width = 4, height = 4)
ggplot(cds2_symbol_results, aes(x = logFC, y = neglog10_p_fdr_adj)) + 
  geom_hline(yintercept = 4, linetype = 2) +
  geom_vline(xintercept = 0.58, linetype = 2) +
  geom_vline(xintercept = -0.58, linetype = 2) +
  geom_point(alpha = 0.15, fill = "black", pch=21, size = 3) +
  theme_classic() +
  geom_point(data = cds2_symbol_results[cds2_symbol_results$symbol == "CIC",], fill = "red", pch=21, size = 3) +
  geom_point(data = cds2_symbol_results[cds2_symbol_results$symbol %in% c("SPRY1", "SPRY2", "SPRY3", "SPRY4"),], fill = "green3", pch=21, size = 3) +
  geom_point(data = cds2_symbol_results[cds2_symbol_results$symbol %in% c("SPRED1", "SPRED2", "SPRED3"),], fill = "palegreen2", pch=21, size = 3) +
  xlab("log2(FC)")+
  ylab("-log10(q)")+
  ggtitle("CDS2, CIC (red), SPRY1/2/3/4 (darker green), SPRED1/2/3 (lighter green)")
dev.off()

#this plot colors CIC red, POLE pink, and CDH tan
pdf(file = "plots/CDS2.POLE.CDH.pdf", width = 4, height = 4)
ggplot(cds2_symbol_results, aes(x = logFC, y = neglog10_p_fdr_adj)) + 
  geom_hline(yintercept = 4, linetype = 2) +
  geom_vline(xintercept = 0.58, linetype = 2) +
  geom_vline(xintercept = -0.58, linetype = 2) +
  geom_point(alpha = 0.15, fill = "black", pch=21, size = 3) +
  theme_classic() +
  geom_point(data = cds2_symbol_results[cds2_symbol_results$symbol == "CIC",], fill = "red", pch=21, size = 3) +
  geom_point(data = cds2_symbol_results[cds2_symbol_results$symbol %in% c("POLE"),], fill = "pink", pch=21, size = 3) +
  geom_point(data = cds2_symbol_results[cds2_symbol_results$symbol %in% c("CDH4"),], fill = "tan", pch=21, size = 3) +
  xlab("log2(FC)")+
  ylab("-log10(q)")+
  ggtitle("CDS2, CIC (red), POLE (pink), CDH4 (tan)")
dev.off()

#extract logCPM values for dimensionality reduction and etc.
cds2_logcpm = cpm(cds2, log = T)
cds2_logcpm_df = as.data.frame(cds2_logcpm)
cds2_logcpm_df = tibble::rownames_to_column(cds2_logcpm_df, "ENSG_ID")

#visualize by PCA using logCPM values
cds2_pca_obj = fast.prcomp(t(cds2_logcpm), center=T, scale=T)
pdf(file = "plots/CDS2_logCPM_PCA.pdf", width = 3, height = 3)
ggbiplot(cds2_pca_obj, var.axes = F, groups = c("siSCRM", "siSCRM", "siCIC", "siCIC"),) +
  theme_classic()
dev.off()













#Making a plot comparing the number of trans-breakpoint reads found by the "grep" command in the fastq files for the data...
#Manually entering the numbers from the data shown in "fusion_search output 032923.png"
#Grouping all four data points per condition (two sets of reads per replicate, with two replicates per condition)
#These numbers are just the result of running wc -l on .txt files output from grep searching the fastq files for the following appropriate 20nt trans-breakpoint sequences:
#X1: CTCTGGGGGTGGAGGCGACG
#CDS2: CGGACTCTGGGGAGCCGGCG
X1_siCIC_fusion_reads = c(12, 19, 22, 24)
X1_siSCRMBL_fusion_reads = c(72, 61, 45, 46)
CDS2_siSCRMBL_fusion_reads = c(35, 33, 40, 35)
CDS2_siCIC_fusion_reads = c(11, 10, 19, 17)

#since this data is just searching raw reads and is completely non-normalized, we should compare this to the total number of raw reads (read1+read2)
#that came off of the sequencer just to make sure those samples didn't have fewer overall reads.
#this data is in the Novogene QC report, I'm just copying it from there. The data (without headers, I was lazy) is in "raw reads.txt".
#I'm providing each replicate as an individual data point within a variable describing the condition
X1_siSCRMBL_total_raw = c(46920778, 43463344)
X1_siCIC_total_raw = c(43315674, 42538112)
CDS2_siSCRMBL_total_raw = c(42075492, 46026406)
CDS2_siCIC_total_raw = c(42573366, 51285386)

fusion_reads = data.frame(X1_siSCRMBL_fusion_reads, X1_siCIC_fusion_reads, CDS2_siSCRMBL_fusion_reads, CDS2_siCIC_fusion_reads)
total_reads = data.frame(X1_siSCRMBL_total_raw, X1_siCIC_total_raw, CDS2_siSCRMBL_total_raw, CDS2_siCIC_total_raw)

#pivot these longer to make them amenable to ggplot-ing
fusion_reads_pivot = pivot_longer(fusion_reads, cols = c(1:4), names_to = "condition",)
total_reads_pivot = pivot_longer(total_reads, cols = c(1:4), names_to = "condition")

#and now let's just make some nice dot plots showing the data, with T tests comparing siCIC vs. siSCRM within each cell line
#I'm having stat_compare_means display the unadjusted p-value because I am not intending to do all of the possible pairwise comparisons.
#I only care about siSCRM vs siCIC within the same cell line. I could have just manually compared the cell lines individually and then added on the 
#P-values myself. This is just the more convenient way to code it.
#Even if you adjusted p-values, the comparisons are still ns/s as appropriate.
compare_means(value ~ condition, data = total_reads_pivot, method = "t.test")
pdf(file = "plots/total_raw_reads_comparison.pdf", width = 5, height = 5)
ggplot(total_reads_pivot, aes(x = factor(condition, level = c("X1_siSCRMBL_total_raw", "X1_siCIC_total_raw", "CDS2_siSCRMBL_total_raw", "CDS2_siCIC_total_raw")), y = value)) + 
  stat_summary(fun = mean, color = "steelblue4" ,fill = "steelblue1", geom = "bar", alpha = 0.25) +
  geom_point(color = "black") + 
  theme_classic() +
  ylab("Raw Reads (read1+read2)")+
  xlab("Condition") +
  theme(axis.text.x = element_text(angle = 45, vjust = 0.75, hjust = 0.75)) +
  stat_compare_means(method = "t.test", comparisons = list(c(1,2),c(3,4)))
dev.off()
  
compare_means(value ~ condition, data = fusion_reads_pivot, method = "t.test")
pdf(file = "plots/fusion_reads_comparison.pdf", width = 5, height = 5)
ggplot(fusion_reads_pivot, aes(x = factor(condition, level = c("X1_siSCRMBL_fusion_reads", "X1_siCIC_fusion_reads", "CDS2_siSCRMBL_fusion_reads", "CDS2_siCIC_fusion_reads")), y = value)) + 
  stat_summary(fun = mean, color = "darkorchid4" ,fill = "darkorchid1", geom = "bar", alpha = 0.25) +
  geom_point(color = "black") + 
  theme_classic() +
  ylab("Matching Trans-Breakpoint Reads")+
  xlab("Condition") +
  theme(axis.text.x = element_text(angle = 45, vjust = 0.75, hjust = 0.75)) +
  stat_compare_means(method = "t.test", comparisons = list(c(1,2),c(3,4)))
dev.off()




### Also interested in a similar comparison as above, but for reads that contain a 20bp sequence from WT CIC that spans across
# where the breakpoint should be in both cell lines. Thus this lets us look at WT CIC expression in the cell lines

#Manually entering the numbers from the data shown in "wt_and_fusion_search_071023"
#Grouping all four data points per condition (two sets of reads per replicate, with two replicates per condition)
#These numbers are just the result of running wc -l on .txt files output from grep searching the fastq files for the following appropriate 20nt WT CIC sequence:
# CGGACTCTGGCACGGCCCAG (same for both cell lines, since the CIC breakpoint is the same for both)

X1_siCIC_WT_reads = c(22, 18, 14, 13)
X1_siSCRMBL_WT_reads = c(45, 41, 34, 32)
CDS2_siSCRMBL_WT_reads = c(25, 24, 27, 22)
CDS2_siCIC_WT_reads = c(11, 9, 13, 14)

#aggregate into one df
wt_reads = data.frame(X1_siCIC_WT_reads, X1_siSCRMBL_WT_reads, CDS2_siSCRMBL_WT_reads, CDS2_siCIC_WT_reads)

#pivot longer to make amenable for plotting
wt_reads_pivot = pivot_longer(wt_reads, cols = c(1:4), names_to = "condition",)

#and plot as per above

compare_means(value ~ condition, data = wt_reads_pivot, method = "t.test")
pdf(file = "plots/wt_reads_comparison.pdf", width = 5, height = 5)
ggplot(wt_reads_pivot, aes(x = factor(condition, level = c("X1_siSCRMBL_WT_reads", "X1_siCIC_WT_reads", "CDS2_siSCRMBL_WT_reads", "CDS2_siCIC_WT_reads")), y = value)) + 
  stat_summary(fun = mean, color = "darkorchid4" ,fill = "darkorchid1", geom = "bar", alpha = 0.25) +
  geom_point(color = "black") + 
  theme_classic() +
  ylab("Matching WT Reads")+
  xlab("Condition") +
  theme(axis.text.x = element_text(angle = 45, vjust = 0.75, hjust = 0.75)) +
  stat_compare_means(method = "t.test", comparisons = list(c(1,2),c(3,4)))
dev.off()





### Using log(CPM) data I also want to do some correlation analysis between replicates, just for QC.
# There's four sets of replicate pairs so I'm going to do this individually for each set.
# Writing a function would've been faster & simpler but I did not do that.

#X1 scramble
x1_scrm_cor = lm(X1_scrm2~X1_scrm1, x1_logcpm_df)
pdf(file = "plots/X1_siSCRM_rep_cors.pdf", width = 5, height = 5)
ggplot(data = x1_logcpm_df, aes(x = X1_scrm1, y = X1_scrm2)) + 
  geom_point(alpha = 0.2) +
  xlab("X1 siSCRMBL1 log(CPM)") +
  ylab("X1 siSCRMBL2 log(CPM)") +
  stat_cor() +
  geom_smooth(method = lm, formula = y~x, color = "salmon") +
  theme_classic() +
  annotate("text", x = 2, y = 12, label = paste("Regression Line: y = ",format(round(coef(x1_scrm_cor)[2],3), nsmall = 3),"x + ",format(round(coef(x1_scrm_cor)[1],3), nsmall=3),sep = ""))
dev.off()

#X1 siCIC
x1_cic_cor = lm(X1_cic2~X1_cic1, x1_logcpm_df)
pdf(file = "plots/X1_siCIC_rep_cors.pdf", width = 5, height = 5)
ggplot(data = x1_logcpm_df, aes(x = X1_cic1, y = X1_cic2)) + 
  geom_point(alpha = 0.2) +
  xlab("X1 siCIC1 log(CPM)") +
  ylab("X1 siCIC2 log(CPM)") +
  stat_cor() +
  geom_smooth(method = lm, formula = y~x, color = "salmon") +
  theme_classic() +
  annotate("text", x = 2, y = 12, label = paste("Regression Line: y = ",format(round(coef(x1_cic_cor)[2],3), nsmall = 3),"x + ",format(round(coef(x1_cic_cor)[1],3), nsmall=3),sep = ""))
dev.off()

#CDS2 siSCRM
cds2_scrm_cor = lm(CDS2_scrm2~CDS2_scrm1, cds2_logcpm_df)
pdf(file = "plots/CDS2_siSCRM_rep_cors.pdf", width = 5, height = 5)
ggplot(data = cds2_logcpm_df, aes(x = CDS2_scrm1, y = CDS2_scrm2)) + 
  geom_point(alpha = 0.2) +
  xlab("CDS2 siSCRM1 log(CPM)") +
  ylab("CDS2 siSCRM2 log(CPM)") +
  stat_cor() +
  geom_smooth(method = lm, formula = y~x, color = "salmon") +
  theme_classic() +
  annotate("text", x = 2, y = 12, label = paste("Regression Line: y = ",format(round(coef(cds2_scrm_cor)[2],3), nsmall = 3),"x + ",format(round(coef(cds2_scrm_cor)[1],3), nsmall=3),sep = ""))
dev.off()

#CDS2 siCIC
cds2_cic_cor = lm(CDS2_cic2~CDS2_cic1, cds2_logcpm_df)
pdf(file = "plots/CDS2_siCIC_rep_cors.pdf", width = 5, height = 5)
ggplot(data = cds2_logcpm_df, aes(x = CDS2_cic1, y = CDS2_cic2)) + 
  geom_point(alpha = 0.2) +
  xlab("CDS2 siCIC1 log(CPM)") +
  ylab("CDS2 siCIC2 log(CPM)") +
  stat_cor() +
  geom_smooth(method = lm, formula = y~x, color = "salmon") +
  theme_classic() +
  annotate("text", x = 2, y = 12, label = paste("Regression Line: y = ",format(round(coef(cds2_cic_cor)[2],3), nsmall = 3),"x + ",format(round(coef(cds2_cic_cor)[1],3), nsmall=3),sep = ""))
dev.off()


###Now I want to compare which hits from the ChIP-seq data (genes annotated to be near high-confidence CIC-DUX4 peaks) also coincide with
#genes whose expression went down with knockdown of CIC (+CIC-DUX4) in this data.

#To do so I first need to get a set of genes from RNA-seq that are significantly downregulated in at least one cell line
#I'm going to use the cutoffs established in the volcano plots... i.e. -log10(q) > 4, log2FC < -0.58.
#Going to do this with genes that I have symbols for, genes without symbols won't be compatible with the comparison anyways

X1_down = x1_symbol_results[x1_symbol_results$logFC < -0.58 & x1_symbol_results$neglog10_p_fdr_adj > 4,]
CDS2_down = cds2_symbol_results[cds2_symbol_results$logFC < -0.58 & cds2_symbol_results$neglog10_p_fdr_adj > 4,]

#Going to get a list of genes which were significantly downregulated in at least one of these two cell lines
down_genes = unique(c(X1_down$symbol, CDS2_down$symbol)) #at least one

#Writing this to a txt file to do easy gene ontology with
write.table(down_genes, file = "genes_down_in_at_least_one_line.txt", sep = "\t", row.names = F, quote = F, col.names = F)

#I used this file as input to DAVID (https://david.ncifcrf.gov/summary.jsp)
#with all default settings & Homo sapiens to produce the output file "RNAseq_down_at_least_one_line_DAVID_051623.txt"

DAVID = fread("RNAseq_down_at_least_one_line_DAVID_051623.txt")

#I also used the file as input to PANTHER (http://geneontology.org/)
#with default settings, biological process, Homo sapiens, to produce the output file "RNAseq_down_at_least_one_line_PANTHER_051623"

PANTHER = fread("RNAseq_down_at_least_one_line_PANTHER_051623.txt", skip = 11)

#let's make some bar plots of some interesting themes from these quickly 

DAVID_selected = DAVID[DAVID$Term %in% c("KW-0235~DNA replication", "KW-0131~Cell cycle", "GO:0006281~DNA repair", "KW-0227~DNA damage"), c("Term", "FDR")]
DAVID_selected = dplyr::mutate(DAVID_selected, algorithm = "DAVID")
colnames(DAVID_selected) = c("Term", "FDR","Algorithm")
DAVID_selected = dplyr::mutate(DAVID_selected, neglog10FDR = -log10(FDR))


PANTHER_selected = PANTHER[PANTHER$`GO biological process complete` %in% c("nucleic acid metabolic process (GO:0090304)", "mitotic cell cycle (GO:0000278)", "DNA replication (GO:0006260)", "DNA-templated DNA replication (GO:0006261)", "DNA damage response (GO:0006974)
"), c(1,8)]
PANTHER_selected = dplyr::mutate(PANTHER_selected, algorithm = "PANTHER")
colnames(PANTHER_selected) = c("Term", "FDR", "Algorithm")
PANTHER_selected = dplyr::mutate(PANTHER_selected, neglog10FDR = -log10(FDR))


GO_selected = rbind(DAVID_selected, PANTHER_selected)

pdf(file = "GOterms.pdf", width = 10, height = 7)
  ggplot(data = GO_selected, aes(x = reorder(Term, neglog10FDR), y = neglog10FDR)) + 
    geom_bar(stat = "identity", aes(fill = Algorithm)) + 
    scale_y_log10() +
    coord_flip() +
    theme_classic() +
    scale_fill_manual(values = c("#006fc3","#fc558f")) +
    xlab("Term") +
    ylab("-log10(FDR)") + 
    theme(axis.text = element_text(size = 20)) +
    theme(axis.title = element_text(size = 20))
  dev.off()


#Now let's read in the genes annotated for ChIP peaks...
setwd("/Volumes/cuyler/ucsf_okimoto_lab/nick_chipseq_project/hg38_alignments/bams/final")
chip_peaks = fread("CIC_IgG_q0.001_DAC_C3_unique_nonblacklisted_chromosomal_chrLabels_trunc_HOMER_unique_genes.txt", header = F)
setwd("/Volumes/cuyler/ucsf_okimoto_lab/nicole_RNAseq/hg38_data")

#Which gene symbols are both annotated for a ChIP peak, and are sig. downregulated in at least one cell line?
intersect = down_genes[down_genes %in% chip_peaks$V1]

#37 genes are both annotated for a peak and downregulated in at least one line.
write.table(intersect, file = "intersect_genes.txt", quote = F, col.names = F, row.names = F)


#Just to have it, also making DFs of genes significantly upregulated in both cell lines
X1_up = x1_symbol_results[x1_symbol_results$logFC > -0.58 & x1_symbol_results$neglog10_p_fdr_adj > 4,]
CDS2_up = cds2_symbol_results[cds2_symbol_results$logFC > -0.58 & cds2_symbol_results$neglog10_p_fdr_adj > 4,]



#We also want to plot the log2(cpm+2) values for SPRED1 expression in X1 cells.
#Need to pivot the data to do this, but first let's subset to just SPRED1.
#The ENSG ID for SPRED1 is ENSG00000166068

x1_spred1 = x1_logcpm_df[x1_logcpm_df$ENSG_ID == "ENSG00000166068",]

#pivot & make a variable to help color by siRNA condition
x1_spred1_pivot = pivot_longer(x1_spred1, cols = c(2:5), values_to = "logcpm", names_to = "sample")
x1_spred1_pivot = dplyr::mutate(x1_spred1_pivot, Condition = ifelse(sample %in% c("X1_scrm1","X1_scrm2"), "Control", "siCIC"))

#plot
pdf(file = "plots/x1_spred1_zoomout.pdf", width = 4, height = 6)
ggplot(data = x1_spred1_pivot, aes(x = Condition, y = logcpm, fill = Condition)) +
  stat_summary(fun = "mean", geom = "bar") +
  geom_point(size = 4) +
  theme_classic() +
  coord_cartesian(ylim = c(0, 10)) +
  scale_fill_manual(values = c("Gray","Pink")) +
  theme(axis.title = element_text(size = 16), axis.text = element_text(size = 14)) +
  ylab("log2(cpm+2)")
dev.off()

pdf(file = "plots/x1_spred1_zoomin.pdf", width = 4, height = 6)
ggplot(data = x1_spred1_pivot, aes(x = Condition, y = logcpm, fill = Condition)) +
  stat_summary(fun = "mean", geom = "bar") +
  geom_point(size = 4) +
  theme_classic() +
  coord_cartesian(ylim = c(7, 8.5)) +
  scale_fill_manual(values = c("Gray","Pink"))+
  theme(axis.title = element_text(size = 16), axis.text = element_text(size = 14))+
  ylab("log2(cpm+2)")
dev.off()
  


